test/test_network.py

```diff
@@ -9,28 +10,84 @@
 class TestNetwork(unittest.TestCase):
 
     def setUp(self):
-        network = canopen.Network()
-        with self.assertLogs():
-            network.add_node(2, SAMPLE_EDS)
-        network.add_node(3, network[2].object_dictionary)
-        self.network = network
+        self.network = canopen.Network()
 
-    def test_add_node(self):
-        node = self.network[2]
-        self.assertIsInstance(node, canopen.Node)
-        self.assertEqual(node.id, 2)
+    def test_network_add_node(self):
+        # Add using str.
+        with self.assertLogs():
+            node = self.network.add_node(2, SAMPLE_EDS)
         self.assertEqual(self.network[2], node)
-        self.assertEqual(len(self.network), 2)
+        self.assertEqual(node.id, 2)
+        self.assertIsInstance(node, canopen.RemoteNode)
+
+        # Add using OD.
+        node = self.network.add_node(3, self.network[2].object_dictionary)
+        self.assertEqual(self.network[3], node)
+        self.assertEqual(node.id, 3)
+        self.assertIsInstance(node, canopen.RemoteNode)
+
+        # Add using RemoteNode.
+        with self.assertLogs():
+            node = canopen.RemoteNode(4, SAMPLE_EDS)
+        self.network.add_node(node)
+        self.assertEqual(self.network[4], node)
+        self.assertEqual(node.id, 4)
+        self.assertIsInstance(node, canopen.RemoteNode)
+
+        # Add using LocalNode.
+        with self.assertLogs():
+            node = canopen.LocalNode(5, SAMPLE_EDS)
+        self.network.add_node(node)
+        self.assertEqual(self.network[5], node)
+        self.assertEqual(node.id, 5)
+        self.assertIsInstance(node, canopen.LocalNode)
+
+        # Verify that we've got the correct number of nodes.
+        self.assertEqual(len(self.network), 4)
+
+    def test_network_add_node_upload_eds(self):
+        # Will err because we're not connected to a real network.
+        with self.assertLogs(level=logging.ERROR):
+            self.network.add_node(2, SAMPLE_EDS, upload_eds=True)
```

Comment:
This test can be enhanced for example by mocking `canopen.sdo.SdoClient.open` to return a pre-filled StringIO buffer.